¡Hola, Tom!

Mi nombre es Tonatiuh Cruz. Me complace revisar tu proyecto hoy.

Al identificar cualquier error inicialmente, simplemente los destacaré. Te animo a localizar y abordar los problemas de forma independiente como parte de tu preparación para un rol como data-scientist. En un entorno profesional, tu líder de equipo seguiría un enfoque similar. Si encuentras la tarea desafiante, proporcionaré una pista más específica en la próxima iteración.

Encontrarás mis comentarios a continuación - **por favor no los muevas, modifiques o elimines**.

Puedes encontrar mis comentarios en cajas verdes, amarillas o rojas como esta:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Éxito. Todo está hecho correctamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Observaciones. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Necesita corrección. El bloque requiere algunas correcciones. El trabajo no puede ser aceptado con comentarios en rojo.
</div>

Puedes responderme utilizando esto:

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class=“tocSkip”></a>
</div>

<div class="alert alert-block alert-success">
<b>Resumen de la revisión 1</b> <a class="tocSkip"></a>

Hola Tom! Has hecho un excelente trabajo con el proyecto, cada vez más cercas de convertirte en un cientifica de datos. Realizaste la carga de bases, su análisis inicial, sus ajustes, un análisis de algunos variables conmplementado con gráficas y el desarrollo de las pruebas d ehipótesis. Solamente te dejo algunos comentarios para complementar el análisis. 

Sigue con el excelente trabajo!
</div>

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante, Iteración 1</b> <a class=“tocSkip”></a>

Hola Tionatiuh, muchas gracias por todos los comentarios, ya los he revisado y he seguido tus consejos al pié de la letra, me han ayudado a mejorar significativamente este proyecto del Sprint 5.
</div>

# 1 ¿Cuál es la mejor tarifa?

Trabajas como analista para el operador de telecomunicaciones Megaline. La empresa ofrece a sus clientes dos tarifas de prepago, Surf y Ultimate. El departamento comercial quiere saber cuál de las tarifas genera más ingresos para poder ajustar el presupuesto de publicidad.

Vas a realizar un análisis preliminar de las tarifas basado en una selección de clientes relativamente pequeña. Tendrás los datos de 500 clientes de Megaline: quiénes son los clientes, de dónde son, qué tarifa usan, así como la cantidad de llamadas que hicieron y los mensajes de texto que enviaron en 2018. Tu trabajo es analizar el comportamiento de los clientes y determinar qué tarifa de prepago genera más ingresos.

# Contenido <a id='back'></a>

* [Introducción](#intro)
    * [Descripción de los datos](#data_review)
    * [Diccionario de datos](#dictionary)
    * [Plan de solucion](#plan)
* [1  ¿Cuál es la mejor tarifa?](#plan)
    * [1.1  Inicialización](#kickoff)
    * [1.2  Cargar datos](#dataload)
    * [1.3  Preparar los datos](#data_preprocessing)
    * [1.4  Tarifas](#fees)
    * [1.5  Corregir datos](#fess_wrangling)
    * [1.6  Enriquecer los datos](#fees_processing)
    * [1.7  Usuarios/as](#users)
        * [1.7.1  Corregir los datos](#users_wrangling)
        * [1.7.2  Enriquecer los datos](#users_processing)
    * [1.8  Llamadas](#calls)
        * [1.8.1  Corregir los datos](#calls_wrangling)
        * [1.8.2  Enriquecer los datos](calls_processing)
    * [1.9  Mensajes](#sms)
        * [1.9.1  Corregir los datos](#sms__wrangling)
        * [1.9.2  Enriquecer los datos](#calls_processing)
    * [1.10  Internet](#data)
        * [1.10.1  Corregir los datos](#data_wrnagling)
        * [1.10.2  Enriquecer los datos](#data_processing)
    * [1.11  Estudiar las condiciones de las tarifas](#fees_analisys)
    * [1.12  Agregar datos por usuario](#data_adding)
    * [1.13  Estudia el comportamiento de usuario](#user_consumption)
        * [1.13.1  Llamadas](#calls_consumption)
        * [1.13.2  Mensajes](#sms_consumption)
        * [1.13.3  Internet](#data_consumption)
    * [1.14  Ingreso](#incomes)
    * [1.15  Prueba las hipótesis estadísticas](#hypothesis)
* [1.16  Conclusión general](#end)

# Introducción  <a id='intro'></a>

La empresa operadora de Telecomunicaciones MEGALINE ofrece a sus clientes dos tarifas de prepago, Surf y Ultimate. Necesitamos saber cuál de los dos planes genera más ingresos para ajustar el presupuesto de publicidad, de acuerdo a los requerimientos del departamento comercial de la compañía.

Vamos a realizar un análisis preliminar de las tarifas basado en una selección de clientes relativamente pequeña. Tendrás los datos de 500 clientes de Megaline: quiénes son los clientes, de dónde son, qué tarifa usan, así como la cantidad de llamadas que hicieron y los mensajes de texto que enviaron en 2018. Tu trabajo es analizar el comportamiento de los clientes y determinar qué tarifa de prepago genera más ingresos. Más adelante, encontrarás en las instrucciones del proyecto cuáles son exactamente los aspectos del comportamiento de los clientes que debes analizar. Determinar qué plan, en promedio, aporta más ingresos es una cuestión que se abordará mediante pruebas estadísticas. Más adelante encontrarás más información al respecto en la sección de instrucciones del proyecto.

Descripción de las tarifas:

Nota: Megaline redondea los segundos a minutos y los megabytes a gigabytes. Para las llamadas, cada llamada individual se redondea: incluso si la llamada duró solo un segundo, se contará como un minuto. Para el tráfico web, las sesiones web individuales no se redondean. En vez de esto, el total del mes se redondea hacia arriba. Si alguien usa 1025 megabytes este mes, se le cobrarán 2 gigabytes.

A continuación, puedes ver una descripción de las tarifas:

Surf

1.	Pago mensual: $20 usd.
2.	500 minutos al mes, 50 SMS y 15 GB de datos.
3.	Si se exceden los límites del paquete:

    o	1 minuto: 3 centavos.

    o	1 SMS: 3 centavos.

    o	1 GB de datos: $10 usd.

Ultimate

1.	Pago mensual: $70 usd.
2.	3000 minutos al mes, 1000 SMS y 30 GB de datos.
3.	Si se exceden los límites del paquete:

    o	1 minuto: 1 centavo.

    o	1 SMS: 1 centavo.
    
    o	1 GB de datos: $7 usd.

[Volver a Contenido](#back)

[Te proporcionamos algunos comentarios para orientarte mientras completas este proyecto. Pero debes asegurarte de eliminar todos los comentarios entre corchetes antes de entregar tu proyecto.]

[Antes de sumergirte en el análisis de datos, explica por tu propia cuenta el propósito del proyecto y las acciones que planeas realizar.]

[Ten en cuenta que estudiar, modificar y analizar datos es un proceso iterativo. Es normal volver a los pasos anteriores y corregirlos/ampliarlos para permitir nuevos pasos.]

# Descripción de los datos <a id='data_review'></a>

Para este trabajo se tienen los siguientes 5 archivos de datos (`/datasets/megaline_users.csv `, `/datasets/megaline_plans.csv `, `/datasets/megaline_calls.csv `, `/datasets/megaline_internet.csv ` y `/datasets/megaline_messages.csv `) los cuales se leeran con la librería Pandas de Python, utilizando el método de lectura de archivos separados por comas `pd.read_csv()` usando el parámetro sep=';' para leer los datos correctamente. 
De primer instancia, se verificará la información para cada uno de los DataFrames.

## Diccionario de datos <a id='dictionary'></a>

Contamos con cinco Datasets en el conjunto de datos, usaremos todas para hacer el preprocesamiento de los datos y el análisis exploratorio de los datos.
A continuación se muestra el diccionario de datos que enumeroa las columnas de cada uno de los Datasets y se describen los datos que contienen.

- `megaline_users.csv`: cada fila corresponde a un usuario en la Compañia Megaline (datos sobre los usuarios).
    - `'user_id'`: identificador único del usuario.
    - `'first_name'`: nombre del usuario.
    - `'last_name'`: apellido del usuario.
    - `'age'`: edad del usuario (en años).
    - `'reg_date'`: fecha de suscripción (dd, mm, aa).
    - `'churn_date'`: la fecha en la que el usuario dejó de usar el servicio (si el valor es ausente, la tarifa se estaba usando cuando fue extraída esta base de datos).
    - `'city'`: ciudad de residencia del usuario.
    - `'plan'`: nombre de la tarifa ('surf' o 'ultimate')

- `megaline_plans.csv `: cada fila corresponde a... (datos sobre las tarifas).
    - `'plan_name'`: nombre de la tarifa.
    - `'usd_monthly_fee'`: pago mensual en dólares estadounidenses.
    - `'minutes_included'`: minutos incluidos al mes.
    - `'messages_included'`: SMS incluidos al mes.
    - `'mb_per_month_included'`: datos incluidos al mes (en megabytes).
    - `'usd_per_minute'`: precio por minuto tras exceder los límites del paquete (por ejemplo, si el paquete incluye 100 minutos, el operador cobrará el minuto 101).
    - `'usd_per_message'`: precio por SMS tras exceder los límites del paquete.
    - `'usd_per_gb'`: precio por gigabyte de los datos extra tras exceder los límites del paquete (1 GB = 1024 megabytes).

- `megaline_calls.csv`: cada fila corresponde a una llamada única que realizaron los clientes (datos sobre las llamadas).
    - `'id'`: identificador único de la llamada.
    - `'call_date'`: fecha de la llamada.
    - `'duration'`: duración de la llamada (en minutos).
    - `'user_id'`: el identificador del usuario que realiza la llamada.

- `megaline_internet.csv`: cada fila corresponde a una consulta en internet que realizaron los clientes (datos sobre las sesiones web).
    - `'id'`: identificador único de la sesión.
    - `'mb_used'`: el volumen de datos gastados durante la sesión (en megabytes).
    - `'session_date'`: fecha de la sesión web.
    - `'user_id'`: identificador del usuario.

- `megaline_messages.csv`: cada fila corresponde a un mensaje tipo sms que realizaron los clientes (datos sobre los SMS).
    - `'id'`: identificador único del SMS.
    - `'message_date'`: fecha del SMS.
    - `'user_id'`: el identificador del usuario que manda el SMS.

[Volver a Contenido](#back)

## Plan de solución <a id='plan1'></a>

Descripción de los datos.

Para iniciar con la revisión de todos los Datasets, vamos a importa las librerias de Pandas, Numpy y Stats de Scipy que nos serviran para crear un Dataframe por cada uno de los archivos csv.

De esta manera será posible revisar la información contenida en cada archivo, realizando el Preprocesamiento de los datos; como corrección de los nombres de las columnas para cumplir con el formato SnakeCase; rellenar valores ausentes, y eliminar datos duplicados de ser necesario.

Finalmente, realizaremos el Análisis de los datos, creando Hipótesis que nos ayuden a tomar buenas desiciones con la información qye tenemos.

## 1.1 Inicialización <a id='kickoff'></a>

In [1]:
# Cargar todas las librerías
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
from scipy import stats as st
import math

<div class="alert alert-block alert-warning">
<b>Comentario revisor</b> <a class="tocSkip"></a>


Recomiendo cargar la libreria math. De esta manera, puedes realizar redondeos con otros métodos. 
</div>

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante, Iteración 1</b> <a class=“tocSkip”></a>

Listo Tionatiuh, ya he agregado y cargado la librería de math para el redondeo de las cifras más adelante, gracias.
</div>

## 1.2 Cargar datos     <a id='dataload'></a>

In [2]:
# Carga los archivos de datos en diferentes DataFrames
df_users = pd.read_csv(r'C:\Users\aldot\OneDrive\TripleTen_DS\Sprint 5_Análisis estadístico de datos\Proyecto SP5\Datasets\megaline_users.csv', sep=',')
df_plans = pd.read_csv(r'C:\Users\aldot\OneDrive\TripleTen_DS\Sprint 5_Análisis estadístico de datos\Proyecto SP5\Datasets\megaline_plans.csv', sep=',')
df_calls = pd.read_csv(r'C:\Users\aldot\OneDrive\TripleTen_DS\Sprint 5_Análisis estadístico de datos\Proyecto SP5\Datasets\megaline_calls.csv', sep=',')
df_data = pd.read_csv(r'C:\Users\aldot\OneDrive\TripleTen_DS\Sprint 5_Análisis estadístico de datos\Proyecto SP5\Datasets\megaline_internet.csv', sep=',')
df_sms = pd.read_csv(r'C:\Users\aldot\OneDrive\TripleTen_DS\Sprint 5_Análisis estadístico de datos\Proyecto SP5\Datasets\megaline_messages.csv', sep=',')

## 1.3 Preparar los datos <a id='data_preprocessing'></a>

[Los datos para este proyecto se dividen en varias tablas. 
Explora cada una para tener una comprensión inicial de los datos. 
Si es necesario, haz las correcciones requeridas en cada tabla.]

## 1.4 Tarifas <a id='fees'></a>

In [3]:
# Imprime la información general/resumida sobre el DataFrame de las tarifas
df_plans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   messages_included      2 non-null      int64  
 1   mb_per_month_included  2 non-null      int64  
 2   minutes_included       2 non-null      int64  
 3   usd_monthly_pay        2 non-null      int64  
 4   usd_per_gb             2 non-null      int64  
 5   usd_per_message        2 non-null      float64
 6   usd_per_minute         2 non-null      float64
 7   plan_name              2 non-null      object 
dtypes: float64(2), int64(5), object(1)
memory usage: 260.0+ bytes


In [4]:
# Imprime una muestra de los datos para las tarifas
df_plans.head()


messages_included  mb_per_month_included  minutes_included  \
0                 50                  15360               500   
1               1000                  30720              3000   

   usd_monthly_pay  usd_per_gb  usd_per_message  usd_per_minute plan_name  
0               20          10             0.03            0.03      surf  
1               70           7             0.01            0.01  ultimate

[Volver a Contenido](#back)

## Descripción

El dataframe de los planes contiene únicamente 2 filas y 8 columnas, con información sobre las tarifas que se ofrecen.

La mayoría de los datos de las columnas son de tipo números enteros o integer (int64), hay dos columnas con información monetaria (usd) que son de tipo números fraccionarios o float (float64), y finalmente una columna con información de tipo cualitativa o descriptiva (object) que son los nombres de los planes.

La información se ve bien pero podemos cambiar los tipos de datos en todas las columnas que contengan información monetaria en usd, para homologar todas las columnas con esta información.

No se tienen datos eliminados ni ausentes, los nombres de los planes están en minusculas, los nombres de las columnas cumplen el formato snake case, nombres en minúsculas, separación con guiones bajos, sin espacios al principio ni al final.

Por lo que no hay mucho que corregir en este primer dataframe revisado.

## 1.5 Corregir datos <a id='fees_wrangling'></a>

[Corrige los problemas obvios con los datos basándote en las observaciones iniciales.]

In [5]:
# Vamos a cambiar los Dtypes en todas las columnas con información monetaria (usd) de int64 a float64:
df_plans['usd_monthly_pay'] = df_plans['usd_monthly_pay'].astype('float64')
df_plans['usd_per_gb'] = df_plans['usd_per_gb'].astype('float64')
df_plans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   messages_included      2 non-null      int64  
 1   mb_per_month_included  2 non-null      int64  
 2   minutes_included       2 non-null      int64  
 3   usd_monthly_pay        2 non-null      float64
 4   usd_per_gb             2 non-null      float64
 5   usd_per_message        2 non-null      float64
 6   usd_per_minute         2 non-null      float64
 7   plan_name              2 non-null      object 
dtypes: float64(4), int64(3), object(1)
memory usage: 260.0+ bytes


[Volver a Contenido](#back)

## 1.6 Enriquecer los datos <a id='fees_processing'></a>

[Agrega factores adicionales a los datos si crees que pudieran ser útiles.]

In [6]:
# Agregamos una columna que contenga el los Mb (Megabytes) incluídos en Gb (Gigabytes)
gb_included = (df_plans['mb_per_month_included'] / 1024).astype('int64')
df_plans.insert(2, 'gb_included', gb_included)

# renombramos las columnas más descriptivas
df_plans.columns = ['sms_included', 'mb_per_month_included', 'gb_included', 'min_included', 'usd_monthly_pay', 'usd_per_gb', 'usd_per_sms', 'usd_per_minute','plan']
df_plans

sms_included  mb_per_month_included  gb_included  min_included  \
0            50                  15360           15           500   
1          1000                  30720           30          3000   

   usd_monthly_pay  usd_per_gb  usd_per_sms  usd_per_minute      plan  
0             20.0        10.0         0.03            0.03      surf  
1             70.0         7.0         0.01            0.01  ultimate


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy buen trabajo!! Es correcto considerar que 1024 megabytes son 1 gigabytes.
    
</div>

[Volver a Contenido](#back)

## 1.7 Usuarios/as <a id='users'></a>

In [7]:
# Imprime la información general/resumida sobre el DataFrame de usuarios
df_users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   user_id     500 non-null    int64 
 1   first_name  500 non-null    object
 2   last_name   500 non-null    object
 3   age         500 non-null    int64 
 4   city        500 non-null    object
 5   reg_date    500 non-null    object
 6   plan        500 non-null    object
 7   churn_date  34 non-null     object
dtypes: int64(2), object(6)
memory usage: 31.4+ KB


In [8]:
# Imprime una muestra de datos para usuarios
df_users.sample(5)

user_id first_name last_name  age  \
348     1348       Carl  Rutledge   60   
289     1289      Steve     Blair   47   
270     1270   Anjelica      Chan   70   
64      1064    Gerardo    Ashley   46   
335     1335       Gail   Coleman   61   

                                                city    reg_date      plan  \
348                  Detroit-Warren-Dearborn, MI MSA  2018-10-05      surf   
289                  Detroit-Warren-Dearborn, MI MSA  2018-11-12  ultimate   
270  Philadelphia-Camden-Wilmington, PA-NJ-DE-MD MSA  2018-09-18  ultimate   
64         New York-Newark-Jersey City, NY-NJ-PA MSA  2018-06-03      surf   
335              Dallas-Fort Worth-Arlington, TX MSA  2018-06-02      surf   

    churn_date  
348        NaN  
289        NaN  
270        NaN  
64         NaN  
335        NaN

En el dataframe de usuarios, podemos ver 8 columnas con diferentes datos, como 'user_id', 'first_name', 'last_name', 'age', 'city', 'reg_date', 'plan', 'churn_date'.

Al revisar los Dtypes podemos ver eu la columna con fecha de registro 'reg_date' y la columna 'churn_date', son de tipo object; lo cual no es lo más correcto y debe ser corregido a tipo de fecha y con formato de fecha.

También en la última columna llamada 'churn_date' solo tenemos 34 registros y los demás son valores ausentes, lo cual debe corregirse para evitar problemas en el análisis posterios, estos podrían sustituirse con un valor cero o desconocido (unknown).

La muestra a analizar es de 500 usuarios.

### 1.7.1 Corregir los datos <a id='users_wrangling'></a>

Es necesario cambiar el formato de las columnas con fechas, para esto utilizaremos un método de Pandas llamado to_datetime.

Tanto la columna llamada 'reg_date' como la columna 'churn_date' se modificará su formato.

Rellenar los valores ausentes en la columna 'churn_date' con el nombre 'unknown'.

Renombraremos las columnas con nombres más descriptivos.

In [9]:
# cambiamos el formato a fecha en las columnas 'reg_date' y 'churn_date'
df_users['reg_date'] = pd.to_datetime(df_users['reg_date'], infer_datetime_format=True)
df_users['churn_date'] = pd.to_datetime(df_users['churn_date'], infer_datetime_format=True)

df_users['churn_date'] = df_users['churn_date'].fillna('Unknown')   # Rellenamos los valores ausentes
df_users.columns = ['user_id', 'first_name', 'last_name', 'user_age', 'user_city', 'reg_date', 'plan', 'churn_date']   # renombramos las columnas más descriptivas

df_users.info()     # verificamos la información general y los Dtypes de las columnas
df_users

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   user_id     500 non-null    int64         
 1   first_name  500 non-null    object        
 2   last_name   500 non-null    object        
 3   user_age    500 non-null    int64         
 4   user_city   500 non-null    object        
 5   reg_date    500 non-null    datetime64[ns]
 6   plan        500 non-null    object        
 7   churn_date  500 non-null    object        
dtypes: datetime64[ns](1), int64(2), object(5)
memory usage: 31.4+ KB


C:\Users\aldot\AppData\Local\Temp\ipykernel_16260\3118405040.py:2: UserWarning: The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.
  df_users['reg_date'] = pd.to_datetime(df_users['reg_date'], infer_datetime_format=True)
C:\Users\aldot\AppData\Local\Temp\ipykernel_16260\3118405040.py:3: UserWarning: The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.
  df_users['churn_date'] = pd.to_datetime(df_users['churn_date'], infer_datetime_format=True)


user_id first_name   last_name  user_age  \
0       1000   Anamaria       Bauer        45   
1       1001     Mickey   Wilkerson        28   
2       1002     Carlee     Hoffman        36   
3       1003   Reynaldo     Jenkins        52   
4       1004    Leonila    Thompson        40   
..       ...        ...         ...       ...   
495     1495      Fidel      Sharpe        67   
496     1496      Ariel    Shepherd        49   
497     1497      Donte     Barrera        49   
498     1498       Scot  Williamson        51   
499     1499      Shena     Dickson        37   

                                     user_city   reg_date      plan churn_date  
0        Atlanta-Sandy Springs-Roswell, GA MSA 2018-12-24  ultimate    Unknown  
1              Seattle-Tacoma-Bellevue, WA MSA 2018-08-13      surf    Unknown  
2         Las Vegas-Henderson-Paradise, NV MSA 2018-10-21      surf    Unknown  
3                                Tulsa, OK MSA 2018-01-28      surf    Unknown  
4              Seattle-Tacoma-Bellevue, WA MSA 2018-05-23      surf    Unknown  
..                                         ...        ...       ...        ...  
495  New York-Newark-Jersey City, NY-NJ-PA MSA 2018-09-04      surf    Unknown  
496               New Orleans-Metairie, LA MSA 2018-02-20      surf    Unknown  
497     Los Angeles-Long Beach-Anaheim, CA MSA 2018-12-10  ultimate    Unknown  
498  New York-Newark-Jersey City, NY-NJ-PA MSA 2018-02-04      surf    Unknown  
499          Orlando-Kissimmee-Sanford, FL MSA 2018-05-06      surf    Unknown  

[500 rows x 8 columns]

### 1.7.2 Enriquecer los datos <a id='users_processing'></a>

Al tener una columna con la Fecha de Registro, podemos dividirla para extraer el mes y el año y agregarlos en nuevas columnas cada valor.
Esta información separada, nos podría ser útil para algun análisis más adelante.

In [10]:
# Separaremos los meses de la fecha de registro
reg_month = df_users['reg_date'].dt.month
df_users.insert(6, 'reg_month', reg_month)   # agregamos una columna con el mes de registro

df_users    # imrpimimos el nuevo datafram

user_id first_name   last_name  user_age  \
0       1000   Anamaria       Bauer        45   
1       1001     Mickey   Wilkerson        28   
2       1002     Carlee     Hoffman        36   
3       1003   Reynaldo     Jenkins        52   
4       1004    Leonila    Thompson        40   
..       ...        ...         ...       ...   
495     1495      Fidel      Sharpe        67   
496     1496      Ariel    Shepherd        49   
497     1497      Donte     Barrera        49   
498     1498       Scot  Williamson        51   
499     1499      Shena     Dickson        37   

                                     user_city   reg_date  reg_month  \
0        Atlanta-Sandy Springs-Roswell, GA MSA 2018-12-24         12   
1              Seattle-Tacoma-Bellevue, WA MSA 2018-08-13          8   
2         Las Vegas-Henderson-Paradise, NV MSA 2018-10-21         10   
3                                Tulsa, OK MSA 2018-01-28          1   
4              Seattle-Tacoma-Bellevue, WA MSA 2018-05-23          5   
..                                         ...        ...        ...   
495  New York-Newark-Jersey City, NY-NJ-PA MSA 2018-09-04          9   
496               New Orleans-Metairie, LA MSA 2018-02-20          2   
497     Los Angeles-Long Beach-Anaheim, CA MSA 2018-12-10         12   
498  New York-Newark-Jersey City, NY-NJ-PA MSA 2018-02-04          2   
499          Orlando-Kissimmee-Sanford, FL MSA 2018-05-06          5   

         plan churn_date  
0    ultimate    Unknown  
1        surf    Unknown  
2        surf    Unknown  
3        surf    Unknown  
4        surf    Unknown  
..        ...        ...  
495      surf    Unknown  
496      surf    Unknown  
497  ultimate    Unknown  
498      surf    Unknown  
499      surf    Unknown  

[500 rows x 9 columns]

[Volver a Contenido](#back)

## 1.8 Llamadas <a id='calls'></a>

In [11]:
# Imprime la información general/resumida sobre el DataFrame de las llamadas
df_calls.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 137735 entries, 0 to 137734
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   id         137735 non-null  object 
 1   user_id    137735 non-null  int64  
 2   call_date  137735 non-null  object 
 3   duration   137735 non-null  float64
dtypes: float64(1), int64(1), object(2)
memory usage: 4.2+ MB


In [12]:
# Imprime una muestra de datos para las llamadas
df_calls.sample(5)

id  user_id   call_date  duration
70922  1252_325     1252  2018-10-13      6.70
1009    1007_30     1007  2018-09-04      6.70
85902  1316_163     1316  2018-10-17      4.92
38073  1142_278     1142  2018-10-26      0.00
92836  1334_204     1334  2018-04-28      0.25

En el dataframe de llamadas, tenemos únicamente 4 columnas: 'id', 'user_id', 'call_date' y 'duration'.

Revisando los Dtypes podemos notar que la información de la columna 'call_date' está como un 'object' lo cual no es correcto, tendremos que corregir este formato. al igual que la duración de las llamadas está en minutos.

Se tiene un registro de 137,735 llamadas, sin datos ausentes o nulos.

### 1.8.1 Corregir los datos <a id='calls_wrangling'></a>

Vamos a cambiar el Dtype de la columna 'call_date' para convertirlo en formato de fecha.

También actualizaremos el nombre de las columnas.

Vamos a redondear la duración de las llamadas para efectos prácticos.

In [13]:
df_calls['call_date'] = pd.to_datetime(df_calls['call_date'], infer_datetime_format=True)       # cambiamos el formato a fecha en la columna de 'call_date'
df_calls.columns = ['call_id', 'user_id', 'call_date', 'duration_min']      # Renombramos las columnas para mejor entendimiento
df_calls.info()    # revisamos la información general nuevamente
df_calls

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 137735 entries, 0 to 137734
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   call_id       137735 non-null  object        
 1   user_id       137735 non-null  int64         
 2   call_date     137735 non-null  datetime64[ns]
 3   duration_min  137735 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 4.2+ MB


C:\Users\aldot\AppData\Local\Temp\ipykernel_16260\3572108234.py:1: UserWarning: The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.
  df_calls['call_date'] = pd.to_datetime(df_calls['call_date'], infer_datetime_format=True)       # cambiamos el formato a fecha en la columna de 'call_date'


call_id  user_id  call_date  duration_min
0        1000_93     1000 2018-12-27          8.52
1       1000_145     1000 2018-12-27         13.66
2       1000_247     1000 2018-12-27         14.48
3       1000_309     1000 2018-12-28          5.76
4       1000_380     1000 2018-12-30          4.22
...          ...      ...        ...           ...
137730  1499_199     1499 2018-11-21          8.72
137731  1499_200     1499 2018-10-20         10.89
137732  1499_201     1499 2018-09-21          8.12
137733  1499_202     1499 2018-10-10          0.37
137734  1499_203     1499 2018-12-29         13.86

[137735 rows x 4 columns]

In [14]:
# Redondeamos hacia arriba la duración de las llamadas
df_calls['duration_min'] = np.array(df_calls['duration_min'])
df_calls['duration_min'] = np.ceil(df_calls['duration_min'])
df_calls

call_id  user_id  call_date  duration_min
0        1000_93     1000 2018-12-27           9.0
1       1000_145     1000 2018-12-27          14.0
2       1000_247     1000 2018-12-27          15.0
3       1000_309     1000 2018-12-28           6.0
4       1000_380     1000 2018-12-30           5.0
...          ...      ...        ...           ...
137730  1499_199     1499 2018-11-21           9.0
137731  1499_200     1499 2018-10-20          11.0
137732  1499_201     1499 2018-09-21           9.0
137733  1499_202     1499 2018-10-10           1.0
137734  1499_203     1499 2018-12-29          14.0

[137735 rows x 4 columns]

### 1.8.2 Enriquecer los datos <a id='calls_processing'></a>

En este dataframe de llamadas podemos hacer varias cosas para enriquecer la información.

Primeramente vamos a separar el mes de la fecha de cada llamada.

Seguido de separar la duración en minutos y segundos.

In [15]:
# Separaremos los meses de la fecha de la llamada
call_month = df_calls['call_date'].dt.month
df_calls.insert(3, 'call_month', call_month)   # agregamos una columna con el mes de registro

# Vamos a convertir los minutos en segundos
# df_calls['duration_sec'] = df_calls['duration'] * 60
# df_calls.columns = ['call_id', 'user_id', 'call_date', 'call_month', 'duration_min', 'duration_sec']

df_calls    # imrpimimos el nuevo datafram

call_id  user_id  call_date  call_month  duration_min
0        1000_93     1000 2018-12-27          12           9.0
1       1000_145     1000 2018-12-27          12          14.0
2       1000_247     1000 2018-12-27          12          15.0
3       1000_309     1000 2018-12-28          12           6.0
4       1000_380     1000 2018-12-30          12           5.0
...          ...      ...        ...         ...           ...
137730  1499_199     1499 2018-11-21          11           9.0
137731  1499_200     1499 2018-10-20          10          11.0
137732  1499_201     1499 2018-09-21           9           9.0
137733  1499_202     1499 2018-10-10          10           1.0
137734  1499_203     1499 2018-12-29          12          14.0

[137735 rows x 5 columns]

In [16]:
# Buscamos llamadas que tengan de cero (0) minutos de duración
df_calls[df_calls['duration_min'] == 0]

call_id  user_id  call_date  call_month  duration_min
19        1001_4     1001 2018-12-05          12           0.0
25       1001_13     1001 2018-12-24          12           0.0
31       1001_30     1001 2018-10-17          10           0.0
32       1001_31     1001 2018-11-03          11           0.0
34       1001_35     1001 2018-12-21          12           0.0
...          ...      ...        ...         ...           ...
137714  1499_183     1499 2018-11-10          11           0.0
137715  1499_184     1499 2018-09-15           9           0.0
137720  1499_189     1499 2018-12-27          12           0.0
137722  1499_191     1499 2018-12-24          12           0.0
137725  1499_194     1499 2018-10-31          10           0.0

[26834 rows x 5 columns]

In [17]:
# Filtramos un nuevo Dataframe que contenga llamadas con duración mayor a cero (0)
df_calls = df_calls[df_calls['duration_min'] > 0]
df_calls.info()
df_calls

<class 'pandas.core.frame.DataFrame'>
Index: 110901 entries, 0 to 137734
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   call_id       110901 non-null  object        
 1   user_id       110901 non-null  int64         
 2   call_date     110901 non-null  datetime64[ns]
 3   call_month    110901 non-null  int32         
 4   duration_min  110901 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int32(1), int64(1), object(1)
memory usage: 4.7+ MB


call_id  user_id  call_date  call_month  duration_min
0        1000_93     1000 2018-12-27          12           9.0
1       1000_145     1000 2018-12-27          12          14.0
2       1000_247     1000 2018-12-27          12          15.0
3       1000_309     1000 2018-12-28          12           6.0
4       1000_380     1000 2018-12-30          12           5.0
...          ...      ...        ...         ...           ...
137730  1499_199     1499 2018-11-21          11           9.0
137731  1499_200     1499 2018-10-20          10          11.0
137732  1499_201     1499 2018-09-21           9           9.0
137733  1499_202     1499 2018-10-10          10           1.0
137734  1499_203     1499 2018-12-29          12          14.0

[110901 rows x 5 columns]

<div class="alert alert-block alert-warning">
<b>Comentario Revisor</b> <a class="tocSkip"></a>

Qué podríamos decir hasta el momento de las llamadas que tienen una duración de 0?
</div>

<div class="alert alert-block alert-info">
<b>Respuesta del , Iteración 1</b> <a class=“tocSkip”></a>

Tionatiuh, he redondeado hacia arriba la duración de las llamadas con el método ceil de math, para no tener decimales.

No creo que fuese necesaria agregar una columna con la duración de las llamadas en segundos, por lo que la comente con #.

Finalmente, he filtrado el dataframe de llamadas dejando sólamente registros con duración mayores que cero.
</div>

[Volver a Contenido](#back)

## 1.9 Mensajes <a id='sms'></a>

In [18]:
# Imprime la información general/resumida sobre el DataFrame de los mensajes
df_sms.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76051 entries, 0 to 76050
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   id            76051 non-null  object
 1   user_id       76051 non-null  int64 
 2   message_date  76051 non-null  object
dtypes: int64(1), object(2)
memory usage: 1.7+ MB


In [19]:
# Imprime una muestra de datos para los mensajes
df_sms.sample(5)

id  user_id message_date
26521   1160_34     1160   2018-04-28
57612    1363_4     1363   2018-11-15
8276   1061_282     1061   2018-05-07
59776  1380_193     1380   2018-10-04
33892  1213_102     1213   2018-12-04

En el dataframde de mensajes tipo "sms" tenemos 3 columnas que se llaman 'id', 'user_id' y 'message_date'.

Las columnas tienen Dtype object, int64 y objet, lo cual no es lo correcto al tener un dato de fecha como objeto, este debe ser corregido.

Se tiene un total de 76, 051 mensajes registrados, sin valores nulos.

### 1.9.1 Corregir los datos <a id='sms_wrangling'></a>

Vamos a cambiar el Dtype de la columna 'message_date' para convertirlo en formato de fecha.

Actualisaremos el nombre de las columnas para mejor identificación del id.

In [20]:
df_sms['message_date'] = pd.to_datetime(df_sms['message_date'], infer_datetime_format=True)       # cambiamos el formato a fecha en la columna de 'message_date'
df_sms.columns = ['sms_id', 'user_id', 'message_date']
df_sms. info()    # revisamos la información general nuevamente

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76051 entries, 0 to 76050
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   sms_id        76051 non-null  object        
 1   user_id       76051 non-null  int64         
 2   message_date  76051 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(1), object(1)
memory usage: 1.7+ MB


C:\Users\aldot\AppData\Local\Temp\ipykernel_16260\402032001.py:1: UserWarning: The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.
  df_sms['message_date'] = pd.to_datetime(df_sms['message_date'], infer_datetime_format=True)       # cambiamos el formato a fecha en la columna de 'message_date'


### 1.9.2 Enriquecer los datos <a id='sms_processing'></a>

En este dataframe de mensajes podemos separar el mes de la fecha de cada mensaje.

In [21]:
df_sms['message_month'] = df_sms['message_date'].dt.month   # agregamos una columna con el mes de registro
df_sms  # revisamos el nuevo dataframe

sms_id  user_id message_date  message_month
0      1000_125     1000   2018-12-27             12
1      1000_160     1000   2018-12-31             12
2      1000_223     1000   2018-12-31             12
3      1000_251     1000   2018-12-27             12
4      1000_255     1000   2018-12-26             12
...         ...      ...          ...            ...
76046  1497_526     1497   2018-12-24             12
76047  1497_536     1497   2018-12-24             12
76048  1497_547     1497   2018-12-31             12
76049  1497_558     1497   2018-12-24             12
76050  1497_613     1497   2018-12-23             12

[76051 rows x 4 columns]

[Volver a Contenido](#back)

## 1.10 Internet <a id='data'></a>

In [22]:
# Imprime la información general/resumida sobre el DataFrame de internet
df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 104825 entries, 0 to 104824
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            104825 non-null  object 
 1   user_id       104825 non-null  int64  
 2   session_date  104825 non-null  object 
 3   mb_used       104825 non-null  float64
dtypes: float64(1), int64(1), object(2)
memory usage: 3.2+ MB


In [23]:
# Imprime una muestra de datos para el tráfico de internet
df_data.sample(5)

id  user_id session_date  mb_used
34068   1155_95     1155   2018-12-15   214.74
93379  1435_145     1435   2018-12-12  1279.90
89166   1411_62     1411   2018-10-24   422.40
79852  1373_131     1373   2018-11-04   374.98
78218  1364_446     1364   2018-11-28   579.04

El dataframe de data sobre internet, se pueden notar 4 columnas principales llamadas 'id', 'user_id', 'session_date' y 'mb_user'.

Las columnas tienen un Dtype de 'object', 'int64', 'object' y 'float64' respectivamente para cada una, al igual que en los otros dataframes, es conveniente corregir el Dtype para la informacion que contenga fechas.

El total de registros de consultas en internet, es de 104,825 líneas de las cuales no hay valores ausentes.

### 1.10.1 Corregir los datos <a id='data_wrangling'></a>

Vamos a cambiar el Dtype de la columna 'session_date' para convertirlo en formato de fecha.

Actualisaremos el nombre de las columnas para mejor identificación del id.

In [24]:
df_data['session_date'] = pd.to_datetime(df_data['session_date'], infer_datetime_format=True)       # cambiamos el formato a fecha en la columna de 'message_date'
df_data.columns = ['data_id', 'user_id', 'session_date', 'mb_used']     # Renombramos las columnas
df_data.info()    # revisamos la información general nuevamente
df_data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 104825 entries, 0 to 104824
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   data_id       104825 non-null  object        
 1   user_id       104825 non-null  int64         
 2   session_date  104825 non-null  datetime64[ns]
 3   mb_used       104825 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 3.2+ MB


C:\Users\aldot\AppData\Local\Temp\ipykernel_16260\475451897.py:1: UserWarning: The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.
  df_data['session_date'] = pd.to_datetime(df_data['session_date'], infer_datetime_format=True)       # cambiamos el formato a fecha en la columna de 'message_date'


data_id  user_id session_date  mb_used
0        1000_13     1000   2018-12-29    89.86
1       1000_204     1000   2018-12-31     0.00
2       1000_379     1000   2018-12-28   660.40
3       1000_413     1000   2018-12-26   270.99
4       1000_442     1000   2018-12-27   880.22
...          ...      ...          ...      ...
104820  1499_215     1499   2018-10-20   218.06
104821  1499_216     1499   2018-12-30   304.72
104822  1499_217     1499   2018-09-22   292.75
104823  1499_218     1499   2018-12-07     0.00
104824  1499_219     1499   2018-12-24   758.31

[104825 rows x 4 columns]

<div class="alert alert-block alert-warning">
<b>Comentario revisor</b> <a class="tocSkip"></a>


Te recomiendo solamente comentar que cuando agrupemos los datos vamos a tener que redondear hacia arriba lo valores de la transformación de mbegabyter a gigabyte, para este ejercicio debemos hacer el redondeo superior dado que se cobra el costo extra en cuanto se pasa de los límites.Por lo que es necesario hacer la transformación de mb a gb y después hacer el redondeo.  
</div>

<div class="alert alert-block alert-info">
<b>Respuesta del , Iteración 1</b> <a class=“tocSkip”></a>

Tionatiuh, he agregado una columna donde se converten los megabytes a gigabytes; así como también, he redondeado hacia arriba ambos valores.
</div>

### 1.10.2 Enriquecer los datos <a id='data_processing'></a>

1.- Vamos a separar el mes y el año de la fecha de cada consulta en internet.

2.- Redondearemos hacia arriba los megabytes usados.

3.- Convertiremos los megabytes a gigabytes.

In [25]:
# Separaremos los meses y los años de la fecha de la llamada
data_month = df_data['session_date'].dt.month
df_data.insert(3, 'session_month', data_month)   # agregamos una columna con el mes de la consulta

# Convertimos los megabytes a gigabytes y los agregamos en otra columna
data_gigas = df_data['mb_used'] / 1024
df_data.insert(5, 'gb_used', data_gigas)

df_data.sample(5)      # imprimimos 5 líneas aleatorias del dataframe para ver cómo quedó

data_id  user_id session_date  session_month  mb_used   gb_used
83353   1389_259     1389   2018-10-18             10   459.93  0.449150
4202     1027_30     1027   2018-09-25              9   326.07  0.318428
13622    1064_70     1064   2018-08-20              8     0.00  0.000000
104516  1498_598     1498   2018-07-09              7     0.00  0.000000
104614    1499_9     1499   2018-12-03             12   136.49  0.133291

In [26]:
# Redondeamos hacia arriba los megabytes usados
df_data['mb_used'] = np.array(df_data['mb_used'])
df_data['mb_used'] = np.ceil(df_data['mb_used'])

# Redondeamos hacia arriba los gigabytes usados
df_data['gb_used'] = np.array(df_data['gb_used'])
df_data['gb_used'] = np.ceil(df_data['gb_used'])

df_data

data_id  user_id session_date  session_month  mb_used  gb_used
0        1000_13     1000   2018-12-29             12     90.0      1.0
1       1000_204     1000   2018-12-31             12      0.0      0.0
2       1000_379     1000   2018-12-28             12    661.0      1.0
3       1000_413     1000   2018-12-26             12    271.0      1.0
4       1000_442     1000   2018-12-27             12    881.0      1.0
...          ...      ...          ...            ...      ...      ...
104820  1499_215     1499   2018-10-20             10    219.0      1.0
104821  1499_216     1499   2018-12-30             12    305.0      1.0
104822  1499_217     1499   2018-09-22              9    293.0      1.0
104823  1499_218     1499   2018-12-07             12      0.0      0.0
104824  1499_219     1499   2018-12-24             12    759.0      1.0

[104825 rows x 6 columns]

[Volver a Contenido](#back)

## 1.11 Estudiar las condiciones de las tarifas <a id='fees_analisys'></a>

[Es sumamente importante entender cómo funcionan las tarifas, cómo se les cobra a los usuarios en función de su plan de suscripción. Así que te sugerimos imprimir la información de la tarifa para ver una vez más sus condiciones.]

### Imprime las condiciones de la tarifa y asegúrate de que te quedan claras
Descripción de las tarifas:

Nota: Megaline redondea los segundos a minutos y los megabytes a gigabytes. Para las llamadas, cada llamada individual se redondea: incluso si la llamada duró solo un segundo, se contará como un minuto. Para el tráfico web, las sesiones web individuales no se redondean. En vez de esto, el total del mes se redondea hacia arriba. Si alguien usa 1025 megabytes este mes, se le cobrarán 2 gigabytes.

A continuación, puedes ver una descripción de las tarifas:

Surf

1.	Pago mensual: $20 usd.
2.	500 minutos al mes, 50 SMS y 15 GB de datos.
3.	Si se exceden los límites del paquete:

    o	1 minuto: 3 centavos.

    o	1 SMS: 3 centavos.

    o	1 GB de datos: $10 usd.

Ultimate

1.	Pago mensual: $70 usd.
2.	3000 minutos al mes, 1000 SMS y 30 GB de datos.
3.	Si se exceden los límites del paquete:

    o	1 minuto: 1 centavo.

    o	1 SMS: 1 centavo.
    
    o	1 GB de datos: $7 usd.

[Volver a Contenido](#back)

## 1.12 Agregar datos por usuario <a id='data_adding'></a>

[Ahora que los datos están limpios, agrega los datos por usuario y por periodo para que solo haya un registro por usuario y por periodo. Esto facilitará mucho el análisis posterior.]

In [27]:
# Calcula el número de llamadas hechas por cada usuario al mes. Guarda el resultado.
month_user_calls = df_calls.groupby(['user_id', 'call_month'])['call_id'].count()
month_user_calls.info()
month_user_calls

<class 'pandas.core.series.Series'>
MultiIndex: 2256 entries, (np.int64(1000), np.int32(12)) to (np.int64(1499), np.int32(12))
Series name: call_id
Non-Null Count  Dtype
--------------  -----
2256 non-null   int64
dtypes: int64(1)
memory usage: 28.1 KB


user_id  call_month
1000     12            16
1001     8             22
         9             38
         10            47
         11            49
                       ..
1498     12            32
1499     9             35
         10            41
         11            39
         12            56
Name: call_id, Length: 2256, dtype: int64


<div class="alert alert-block alert-warning">

<b>Comentario revisor</b> <a class="tocSkip"></a>

Si bien es correcta la forma de agrupar para sacar la duración de las llamadas por cada ususario, para proximas ocasiones puedes hacer uso de la siguiente forma:
    
    pivot_calls = calls.pivot_table(index=['user_id', 'month'],
                                values=['duration'],
                                aggfunc=['sum','count']).reset_index()

</div>

In [28]:
# Calcula la cantidad de minutos usados por cada usuario al mes. Guarda el resultado.
month_user_minutes = df_calls.groupby(['user_id', 'call_month'])['duration_min'].sum()
month_user_minutes.info()
month_user_minutes

<class 'pandas.core.series.Series'>
MultiIndex: 2256 entries, (np.int64(1000), np.int32(12)) to (np.int64(1499), np.int32(12))
Series name: duration_min
Non-Null Count  Dtype  
--------------  -----  
2256 non-null   float64
dtypes: float64(1)
memory usage: 28.1 KB


user_id  call_month
1000     12            124.0
1001     8             182.0
         9             315.0
         10            393.0
         11            426.0
                       ...  
1498     12            339.0
1499     9             346.0
         10            385.0
         11            308.0
         12            496.0
Name: duration_min, Length: 2256, dtype: float64

In [29]:
# Calcula el número de mensajes enviados por cada usuario al mes. Guarda el resultado.
month_user_sms = df_sms.groupby(['user_id', 'message_month'])['sms_id'].count()
month_user_sms.info()
month_user_sms

<class 'pandas.core.series.Series'>
MultiIndex: 1806 entries, (np.int64(1000), np.int32(12)) to (np.int64(1497), np.int32(12))
Series name: sms_id
Non-Null Count  Dtype
--------------  -----
1806 non-null   int64
dtypes: int64(1)
memory usage: 22.7 KB


user_id  message_month
1000     12               11
1001     8                30
         9                44
         10               53
         11               36
                          ..
1496     9                21
         10               18
         11               13
         12               11
1497     12               50
Name: sms_id, Length: 1806, dtype: int64

In [30]:
# Calcula el volumen del tráfico de Internet usado por cada usuario al mes. Guarda el resultado.
month_user_data = df_data.groupby(['user_id', 'session_month'])['gb_used'].sum()
month_user_data.info()
month_user_data

<class 'pandas.core.series.Series'>
MultiIndex: 2277 entries, (np.int64(1000), np.int32(12)) to (np.int64(1499), np.int32(12))
Series name: gb_used
Non-Null Count  Dtype  
--------------  -----  
2277 non-null   float64
dtypes: float64(1)
memory usage: 28.4 KB


user_id  session_month
1000     12                4.0
1001     8                21.0
         9                41.0
         10               50.0
         11               49.0
                          ... 
1498     12               56.0
1499     9                37.0
         10               52.0
         11               43.0
         12               58.0
Name: gb_used, Length: 2277, dtype: float64

[Junta los datos agregados en un DataFrame para que haya un registro que represente lo que consumió un usuario único en un mes determinado.]

In [31]:
# Fusiona los datos de llamadas, minutos, mensajes e Internet con base en user_id y month
tripleplay = pd.concat([month_user_calls, month_user_minutes, month_user_sms, month_user_data], axis=1).reset_index()
tripleplay.columns = ['user_id', 'month', 'calls', 'mins', 'sms', 'gb']    # Actualizamos los nombres de las columnas para mejor entendimiento

# Reemplazamos los n[umeros por los nombres de los meses
months = {1:'jan', 2:'feb', 3:'mar', 4:'apr', 5:'may', 6:'jun', 7:'jul', 8:'aug', 9:'sep', 10:'oct', 11:'nov', 12:'dec'}
tripleplay['month'] = tripleplay['month'].replace(months)

# Modificamos los Dtypes para corregirlos
tripleplay['user_id'] = tripleplay['user_id'].astype('object')
# tripleplay['calls'] = tripleplay['calls'].astype('int64')
# tripleplay['minutes'] = tripleplay['minutes'].astype('int64')
# tripleplay['sms'] = tripleplay['sms'].astype('int64')
# tripleplay['gb'] = tripleplay['gb'].astype('int64')

# Rellenamos valores ausentes con cero (0)
tripleplay = tripleplay.fillna(0)

tripleplay.info()   # Imprimimos información general y el nuevo dataframe
tripleplay

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2293 entries, 0 to 2292
Data columns (total 6 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   user_id  2293 non-null   int64  
 1   month    2293 non-null   object 
 2   calls    2293 non-null   float64
 3   mins     2293 non-null   float64
 4   sms      2293 non-null   float64
 5   gb       2293 non-null   float64
dtypes: float64(4), int64(1), object(1)
memory usage: 107.6+ KB


C:\Users\aldot\AppData\Local\Temp\ipykernel_16260\960883874.py:17: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  tripleplay = tripleplay.fillna(0)


user_id month  calls   mins   sms    gb
0        1000   dec   16.0  124.0  11.0   4.0
1        1001   aug   22.0  182.0  30.0  21.0
2        1001   sep   38.0  315.0  44.0  41.0
3        1001   oct   47.0  393.0  53.0  50.0
4        1001   nov   49.0  426.0  36.0  49.0
...       ...   ...    ...    ...   ...   ...
2288     1407   nov    0.0    0.0   1.0   1.0
2289     1482   oct    0.0    0.0   2.0   0.0
2290     1094   oct    0.0    0.0   0.0   3.0
2291     1108   dec    0.0    0.0   0.0   2.0
2292     1311   jun    0.0    0.0   0.0   7.0

[2293 rows x 6 columns]

In [32]:
# Añade la información de la tarifa
tripleplay_plan = tripleplay.merge(df_users, on='user_id')    # combinamos los dataframes de 'all_services' con 'df_users' por el número de id
tripleplay_plan.drop(['first_name', 'last_name', 'user_age', 'user_city', 'reg_date', 'reg_month', 'churn_date'], axis='columns', inplace=True)   # Eliminamos las columnas inecesarias
tripleplay_fee = tripleplay_plan.merge(df_plans, on='plan')    # combinamos el nuevo dataframe con 'df_plans' para agregar la información de las tarifas
tripleplay_fee.info()
tripleplay_fee.head()  # imprimimos una muestra de 10 lineas para revisar

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2293 entries, 0 to 2292
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   user_id                2293 non-null   int64  
 1   month                  2293 non-null   object 
 2   calls                  2293 non-null   float64
 3   mins                   2293 non-null   float64
 4   sms                    2293 non-null   float64
 5   gb                     2293 non-null   float64
 6   plan                   2293 non-null   object 
 7   sms_included           2293 non-null   int64  
 8   mb_per_month_included  2293 non-null   int64  
 9   gb_included            2293 non-null   int64  
 10  min_included           2293 non-null   int64  
 11  usd_monthly_pay        2293 non-null   float64
 12  usd_per_gb             2293 non-null   float64
 13  usd_per_sms            2293 non-null   float64
 14  usd_per_minute         2293 non-null   float64
dtypes: f

user_id month  calls   mins   sms    gb      plan  sms_included  \
0     1000   dec   16.0  124.0  11.0   4.0  ultimate          1000   
1     1001   aug   22.0  182.0  30.0  21.0      surf            50   
2     1001   sep   38.0  315.0  44.0  41.0      surf            50   
3     1001   oct   47.0  393.0  53.0  50.0      surf            50   
4     1001   nov   49.0  426.0  36.0  49.0      surf            50   

   mb_per_month_included  gb_included  min_included  usd_monthly_pay  \
0                  30720           30          3000             70.0   
1                  15360           15           500             20.0   
2                  15360           15           500             20.0   
3                  15360           15           500             20.0   
4                  15360           15           500             20.0   

   usd_per_gb  usd_per_sms  usd_per_minute  
0         7.0         0.01            0.01  
1        10.0         0.03            0.03  
2        10.0         0.03            0.03  
3        10.0         0.03            0.03  
4        10.0         0.03            0.03

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy buen trabajo!! el merge es correcto para unir todas las bases trabajadas.
    
</div>

[Calcula los ingresos mensuales por usuario (resta el límite del paquete gratuito del número total de llamadas, mensajes de texto y datos; multiplica el resultado por el valor del plan de llamadas; añade la tarifa mensual en función del plan de llamadas). Nota: Dadas las condiciones del plan, ¡esto podría no ser tan trivial como un par de líneas! Así que no pasa nada si dedicas algo de tiempo a ello.]

<div class="alert alert-block alert-info">
<b>Respuesta del , Iteración 1</b> <a class=“tocSkip”></a>

Tionatiuh, para el ejercicio de calcular los ingresos mensuales por usuario, he creado algunas funciones con ayuda en el co-learning.
Estoy seguro que la información ahora es correcta, aparte de que se han redondeado las cifras como me lo habías solicitado.
Favor de revisar y hacerme saber cualquier detalle.
Gracias!
</div>

In [33]:
# Echamos de nuevo un vistazo al Dataframe de tarifas por plan
df_plans

sms_included  mb_per_month_included  gb_included  min_included  \
0            50                  15360           15           500   
1          1000                  30720           30          3000   

   usd_monthly_pay  usd_per_gb  usd_per_sms  usd_per_minute      plan  
0             20.0        10.0         0.03            0.03      surf  
1             70.0         7.0         0.01            0.01  ultimate

In [34]:
# Definimos todos los datos de ambos planes 'surf' y 'ultimate' en variables

surf_charge = 20
min_surf = 500
sms_surf = 50
gb_surf = 15
cost_call_surf = 0.03
cost_sms_surf = 0.03
cost_gb_surf = 10

ultimate_charge = 70
min_ultimate = 3000
sms_ultimate = 1000
gb_ultimate = 30
cost_call_ultimate = 0.01
cost_sms_ultimate = 0.01
cost_gb_ultimate = 7

In [35]:
# Función para calcular los minutos extra en las llamadas por usuario

def extra_calls (row):
    plan = row['plan']
    min_used = row['mins']
    
    if plan == 'surf':
        if min_used > min_surf:
            extra_mins = min_used - min_surf
        else:
            extra_mins = 0
    
    if plan == 'ultimate':
        if min_used > min_ultimate:
            extra_mins = min_used - min_ultimate
        else:
            extra_mins = 0
            
    return extra_mins

In [36]:
tripleplay_fee.apply(extra_calls, axis=1)   # Probamos la función para calcular los minutos extra, llamada 'extra_calls'

0       0.0
1       0.0
2       0.0
3       0.0
4       0.0
       ... 
2288    0.0
2289    0.0
2290    0.0
2291    0.0
2292    0.0
Length: 2293, dtype: float64

In [37]:
# Función para calcular los mensajes extra enviados por usuario

def extra_messages (row):
    plan = row['plan']
    sms_sent = row['sms']
    
    if plan == 'surf':
        if sms_sent > sms_surf:
            extra_sms = sms_sent - sms_surf
        else:
            extra_sms = 0
    
    if plan == 'ultimate':
        if sms_sent > sms_ultimate:
            extra_sms = sms_sent - sms_ultimate
        else:
            extra_sms = 0
    
    return extra_sms

In [38]:
tripleplay_fee.apply(extra_messages, axis=1)   # Probamos la función para calcular los mensajes extra enviados, llamada 'extra_sms'

0       0.0
1       0.0
2       0.0
3       3.0
4       0.0
       ... 
2288    0.0
2289    0.0
2290    0.0
2291    0.0
2292    0.0
Length: 2293, dtype: float64

In [39]:
# Función para calcular los gigabytes extra consumidos por usuario

def extra_data (row):
    plan = row['plan']
    data_used = row['gb']
    
    if plan == 'surf':
        if data_used > gb_surf:
            extra_gb = data_used - gb_surf
        else:
            extra_gb =0
            
    if plan == 'ultimate':
        if data_used > gb_ultimate:
            extra_gb = data_used - gb_ultimate
        else:
            extra_gb = 0
    
    return extra_gb

In [40]:
tripleplay_fee.apply(extra_data, axis=1)   # Probamos la función para calcular los gigabytes extra consumidos, llamada 'extra_gb'

0        0.0
1        6.0
2       26.0
3       35.0
4       34.0
        ... 
2288     0.0
2289     0.0
2290     0.0
2291     0.0
2292     0.0
Length: 2293, dtype: float64

Ahora que tenemos las funciones para calcular los servicios extra de acuerdo a las tarifas de los dos planes, vamos a crean un nuevo dataframe con 3 columnas extras para cada excedente de minutos (mins), mensajes (sms) y gigabytes (gb).

In [ ]:
tripleplay_fee['extra_min'] = tripleplay_fee.apply(extra_calls, axis=1)         # Creamos una columna con los minutos extra
tripleplay_fee['extra_sms'] = tripleplay_fee.apply(extra_messages, axis=1)      # Creamos una columna con los mensajes extra
tripleplay_fee['extra_gb'] = tripleplay_fee.apply(extra_data, axis=1)           # Creamos una columna con los gigas extra

tripleplay_extra = tripleplay_fee       # Renombramos el Dataframe y revisamos su nueva información
tripleplay_extra.info()
tripleplay_extra.sample(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2293 entries, 0 to 2292
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   user_id                2293 non-null   int64  
 1   month                  2293 non-null   object 
 2   calls                  2293 non-null   float64
 3   mins                   2293 non-null   float64
 4   sms                    2293 non-null   float64
 5   gb                     2293 non-null   float64
 6   plan                   2293 non-null   object 
 7   sms_included           2293 non-null   int64  
 8   mb_per_month_included  2293 non-null   int64  
 9   gb_included            2293 non-null   int64  
 10  min_included           2293 non-null   int64  
 11  usd_monthly_pay        2293 non-null   float64
 12  usd_per_gb             2293 non-null   float64
 13  usd_per_sms            2293 non-null   float64
 14  usd_per_minute         2293 non-null   float64
 15  extr

user_id month  calls   mins   sms    gb      plan  sms_included  \
572      1123   oct   72.0  620.0  27.0  60.0      surf            50   
2149     1475   jul   14.0  128.0  17.0  11.0  ultimate          1000   
1153     1251   oct   77.0  652.0  33.0  53.0      surf            50   
1940     1422   oct   17.0  164.0   8.0  17.0  ultimate          1000   
721      1155   nov   29.0  277.0  28.0  39.0  ultimate          1000   
5        1001   dec   46.0  412.0  44.0  55.0      surf            50   
1272     1277   sep   93.0  824.0   0.0  52.0      surf            50   
1997     1438   sep   13.0   76.0  20.0  16.0      surf            50   
1178     1256   oct   61.0  531.0  20.0  29.0      surf            50   
1836     1402   sep   21.0  156.0   0.0  26.0      surf            50   

      mb_per_month_included  gb_included  min_included  usd_monthly_pay  \
572                   15360           15           500             20.0   
2149                  30720           30          3000             70.0   
1153                  15360           15           500             20.0   
1940                  30720           30          3000             70.0   
721                   30720           30          3000             70.0   
5                     15360           15           500             20.0   
1272                  15360           15           500             20.0   
1997                  15360           15           500             20.0   
1178                  15360           15           500             20.0   
1836                  15360           15           500             20.0   

      usd_per_gb  usd_per_sms  usd_per_minute  extra_min  extra_sms  extra_gb  
572         10.0         0.03            0.03      120.0        0.0      45.0  
2149         7.0         0.01            0.01        0.0        0.0       0.0  
1153        10.0         0.03            0.03      152.0        0.0      38.0  
1940         7.0         0.01            0.01        0.0        0.0       0.0  
721          7.0         0.01            0.01        0.0        0.0       9.0  
5           10.0         0.03            0.03        0.0        0.0      40.0  
1272        10.0         0.03            0.03      324.0        0.0      37.0  
1997        10.0         0.03            0.03        0.0        0.0       1.0  
1178        10.0         0.03            0.03       31.0        0.0      14.0  
1836        10.0         0.03            0.03        0.0        0.0      11.0

Ahora calcularemos la ganancia sumando el costo del plan más el costo de los servicios extra que se cobran aparte de de la tarifa de los planes.

In [43]:
# Esta función calcula el monto del cargo extra y suma los costos para obtener la ganancia total.

def month_charge (row):
    plan = row['plan']
    extra_min = row['extra_min']
    extra_sms = row['extra_sms']
    extra_gb = row['extra_gb']
    
    if plan == 'surf':
        revenue = surf_charge + ((extra_min * cost_call_surf) + (extra_sms * cost_sms_surf) + (extra_gb * cost_gb_surf))
    else:
        revenue = ultimate_charge + ((extra_min * cost_call_ultimate) + (extra_sms * cost_sms_ultimate) + (extra_gb * cost_gb_ultimate))
        
    return revenue

In [44]:
# Creamos una nueva columna que contenga la ganancia total y aplicamos la función de 'month_charge'

tripleplay_extra['month_revenue'] = tripleplay_extra.apply(month_charge, axis=1)
tripleplay_extra.info()
tripleplay_extra.sample(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2293 entries, 0 to 2292
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   user_id                2293 non-null   int64  
 1   month                  2293 non-null   object 
 2   calls                  2293 non-null   float64
 3   mins                   2293 non-null   float64
 4   sms                    2293 non-null   float64
 5   gb                     2293 non-null   float64
 6   plan                   2293 non-null   object 
 7   sms_included           2293 non-null   int64  
 8   mb_per_month_included  2293 non-null   int64  
 9   gb_included            2293 non-null   int64  
 10  min_included           2293 non-null   int64  
 11  usd_monthly_pay        2293 non-null   float64
 12  usd_per_gb             2293 non-null   float64
 13  usd_per_sms            2293 non-null   float64
 14  usd_per_minute         2293 non-null   float64
 15  extr

user_id month  calls   mins    sms    gb      plan  sms_included  \
795      1171   apr   35.0  369.0    0.0  61.0      surf            50   
183      1046   jul   66.0  509.0   51.0  57.0      surf            50   
1850     1404   aug   52.0  473.0    8.0  36.0      surf            50   
2132     1470   oct   54.0  459.0  145.0   0.0      surf            50   
1904     1414   sep   46.0  458.0   63.0  72.0      surf            50   
390      1082   may   85.0  705.0   28.0  30.0      surf            50   
967      1209   aug    9.0   83.0    4.0   2.0      surf            50   
372      1078   nov   13.0   94.0   36.0  22.0      surf            50   
784      1167   oct   69.0  640.0   52.0  41.0      surf            50   
771      1164   dec   36.0  350.0   26.0  41.0  ultimate          1000   

      mb_per_month_included  gb_included  min_included  usd_monthly_pay  \
795                   15360           15           500             20.0   
183                   15360           15           500             20.0   
1850                  15360           15           500             20.0   
2132                  15360           15           500             20.0   
1904                  15360           15           500             20.0   
390                   15360           15           500             20.0   
967                   15360           15           500             20.0   
372                   15360           15           500             20.0   
784                   15360           15           500             20.0   
771                   30720           30          3000             70.0   

      usd_per_gb  usd_per_sms  usd_per_minute  extra_min  extra_sms  extra_gb  \
795         10.0         0.03            0.03        0.0        0.0      46.0   
183         10.0         0.03            0.03        9.0        1.0      42.0   
1850        10.0         0.03            0.03        0.0        0.0      21.0   
2132        10.0         0.03            0.03        0.0       95.0       0.0   
1904        10.0         0.03            0.03        0.0       13.0      57.0   
390         10.0         0.03            0.03      205.0        0.0      15.0   
967         10.0         0.03            0.03        0.0        0.0       0.0   
372         10.0         0.03            0.03        0.0        0.0       7.0   
784         10.0         0.03            0.03      140.0        2.0      26.0   
771          7.0         0.01            0.01        0.0        0.0      11.0   

      month_revenue  
795          480.00  
183          440.30  
1850         230.00  
2132          22.85  
1904         590.39  
390          176.15  
967           20.00  
372           90.00  
784          284.26  
771          147.00

[Volver a Contenido](#back)

<div class="alert alert-block alert-danger">
<b>Comentario revisor</b> <a class="tocSkip"></a>


Recuerda que para este punto, es necesario que ya tengamos los valores redondeados de la duración de las llamadas y redondear la transformación de megabytes a gigabytes. Esto cobra importancia dado que estamos calculando por cuanto se paso cada usuario, entonces el no redondear puede alterar este valor.
</div>

<div class="alert alert-block alert-info">
<b>Respuesta del , Iteración 1</b> <a class=“tocSkip”></a>

Tionatiuh, anteriormente ya he redondeado hacia arriba todos las cifras, por lo que no debería alterar los resultados; favor de revisar.
Gracias.
</div>

## 1.13 Estudia el comportamiento de usuario <a id='user_consumption'></a>

[Calcula algunas estadísticas descriptivas para los datos agregados y fusionados que nos sean útiles y que muestren un panorama general captado por los datos. Dibuja gráficos útiles para facilitar la comprensión. Dado que la tarea principal es comparar las tarifas y decidir cuál es más rentable, las estadísticas y gráficas deben calcularse por tarifa.]

[En los comentarios hallarás pistas relevantes para las llamadas, pero no las hay para los mensajes e Internet. Sin embargo, el principio del estudio estadístico que se aplica para ellos es el mismo que para las llamadas.]

### 1.13.1 Llamadas <a id='calls_consumption'></a>

In [45]:
# Compara la duración promedio de llamadas por cada plan y por cada mes. 
calls_mean = tripleplay_extra.groupby(['month', 'plan'])['calls'].mean().unstack().reset_index()
calls_mean

plan month       surf   ultimate
0      apr  39.700000  39.047619
1      aug  46.172840  48.000000
2      dec  52.808176  52.139073
3      feb  33.111111  47.285714
4      jan  24.000000  21.250000
5      jul  51.033058  47.355932
6      jun  48.195876  42.148936
7      mar  37.869565  36.583333
8      may  44.727273  45.965517
9      nov  47.791519  49.952756
10     oct  48.578059  51.018868
11     sep  46.752577  48.930233

In [ ]:
# Traza un gráfico de barras para visualizarla.
calls_mean.plot(kind='bar', x='month', xlabel='Mes', ylabel='Llamadas Promedio', title='Llamadas Surf vs Ultimate')

In [ ]:
# Compara el número de minutos mensuales que necesitan los usuarios de cada plan. Traza un histograma.
minutes_plan = tripleplay_plan.groupby('plan_name')['minutes'].plot(kind='hist', bins=20, alpha=0.7, title='Cantidad de Minutos', legend=True)

[Calcula la media y la variable de la duración de las llamadas para averiguar si los usuarios de los distintos planes se comportan de forma diferente al realizar sus llamadas.]

In [ ]:
# Calcula la media y la varianza de la duración mensual de llamadas.
median_calls = tripleplay_plan.groupby(['month'])['calls'].mean().reset_index()
variace_calls = tripleplay_plan.groupby(['month'])['calls'].var().reset_index()

# Renombramos las columnas
median_calls.columns = ['month', 'mean']
variace_calls.columns = ['month', 'variance']

# Imprimimos la media y la varianza
print(median_calls)
print()
print(variace_calls)

In [ ]:
# Traza un diagrama de caja para visualizar la distribución de la duración mensual de llamadas
sns.boxplot(calls_mean)


[Elabora las conclusiones sobre el comportamiento de los usuarios con respecto a las llamadas. ¿Su comportamiento varía en función del plan?]

[Volver a Contenido](#back)

### 1.13.2 Mensajes <a id='sms_consumption'></a>

In [ ]:
# Comprara el número de mensajes que tienden a enviar cada mes los usuarios de cada plan
sms_messages = tripleplay_plan.groupby(['month', 'plan_name'])['sms'].sum().unstack().reset_index()
sms_messages


In [ ]:
# Traza un gráfico de barras para visualizarlo.
sms_messages.plot(kind='bar', x='month', xlabel='Mes', ylabel='Llamadas Promedio', title='Mensajes Surf vs Ultimate')

In [ ]:
# Compara el número de mensajes mensuales que necesitan los usuarios de cada plan. Traza un histograma.
sms_messages = tripleplay_plan.groupby('plan_name')['sms'].plot(kind='hist', bins=20, alpha=0.7, title='Cantidad de Mensajes', legend=True)

In [ ]:
# Calcula la media y la varianza de la duración mensual de llamadas.
median_sms = tripleplay_plan.groupby(['month'])['sms'].mean().reset_index()
variace_sms = tripleplay_plan.groupby(['month'])['sms'].var().reset_index()

# Renombramos las columnas
median_sms.columns = ['month', 'mean']
variace_sms.columns = ['month', 'variance']

# Imprimimos la media y la varianza
print(median_sms)
print()
print(variace_sms)

In [ ]:
# Traza un diagrama de caja para visualizar la distribución de la cantidad mensual de mensajes
sns.boxplot(sms_messages)

[Elabora las conclusiones sobre el comportamiento de los usuarios con respecto a los mensajes. ¿Su comportamiento varía en función del plan?]

[Volver a Contenido](#back)

### 1.13.3 Internet <a id='data_consumption'></a>

In [ ]:
# Compara la cantidad de tráfico de Internet consumido por usuarios por plan
internet_traffic = tripleplay_plan.groupby(['month', 'plan_name'])['mb'].sum().unstack().reset_index()
internet_traffic

In [ ]:
# Traza un gráfico de barras para visualizarlo.
internet_traffic.plot(kind='bar', x='month', xlabel='Mes', ylabel='Megabytes', title='Megabytes Surf vs Ultimate')

In [ ]:
# Compara el número de megabytes mensuales que necesitan los usuarios de cada plan. Traza un histograma.
internet_traffic = tripleplay_plan.groupby('plan_name')['mb'].plot(kind='hist', bins=20, alpha=0.7, title='Cantidad de Megabytes', legend=True)

In [ ]:
# Calcula la media y la varianza de la duración mensual de llamadas.
median_mb = tripleplay_plan.groupby(['month'])['mb'].mean().reset_index()
variace_mb = tripleplay_plan.groupby(['month'])['mb'].var().reset_index()

# Renombramos las columnas
median_mb.columns = ['month', 'mean']
variace_mb.columns = ['month', 'variance']

# Imprimimos la media y la varianza
print(median_mb)
print()
print(variace_mb)

In [ ]:
# Traza un diagrama de caja para visualizar la distribución de la cantidad mensual de tráfico de internet
sns.boxplot(internet_traffic)

[Elabora las conclusiones sobre cómo los usuarios tienden a consumir el tráfico de Internet. ¿Su comportamiento varía en función del plan?]

<div class="alert alert-block alert-success">
<b>Comentario revisor</b> <a class="tocSkip"></a>

 Muy buena prática la de usar distintos tipos de gráficas identificar algunos hallazgos y llegar a conclusiones. En este caso las gráficas de box-plots nos ayudan a identificar outiers que pueden sesgar nuestros resultados, especificamente las pruebas de hipótesis. Es por ello que con estás gráficas puedes complementar el análisis eliminando los valores atipicos. 
</div>

[Volver a Contenido](#back)

## 1.14 Ingreso <a id='incomes'></a>

[Del mismo modo que has estudiado el comportamiento de los usuarios, describe estadísticamente los ingresos de los planes.]

<div class="alert alert-block alert-danger">
<b>Comentario revisor</b> <a class="tocSkip"></a>

Hola, Tom! Has realizado une excelente trabajo con el proyecto. Ya solamente es necesario dearrollar el análisis de la variable ingreso y las pruebas de hipótesis. Recuerda que cualquier pregunta me la puedes dejar para respodnerla

[Elabora las conclusiones sobre cómo difiere el ingreso entre los planes.]

[Volver a Contenido](#back)

## 1.15 Prueba las hipótesis estadísticas <a id='hypotesis'></a>

[Prueba la hipótesis de que son diferentes los ingresos promedio procedentes de los usuarios de los planes de llamada Ultimate y Surf.]

[Elabora las hipótesis nula y alternativa, escoge la prueba estadística, determina el valor alfa.]

In [ ]:
# Prueba las hipótesis



[Prueba la hipótesis de que el ingreso promedio de los usuarios del área NY-NJ es diferente al de los usuarios de otras regiones.]

[Elabora las hipótesis nula y alternativa, escoge la prueba estadística, determina el valor alfa.]

In [ ]:
# Prueba las hipótesis



[Volver a Contenido](#back)

## Conclusión general <a id='end'></a>

[En esta sección final, enumera tus conclusiones importantes. Asegúrate de que estas abarquen todas las decisiones (suposiciones) importantes que adoptaste y que determinaron la forma elegida para procesar y analizar los datos.]